ngraph_bridge/ngraph_utils.cc

```diff
@@ -463,6 +475,198 @@ bool IsProcessedByNgraphPass(Graph* g) {
+  tensorflow::Env* env = Env::Default();
+
+  event.set_wall_time(env->NowMicros() / 1e6);
+  event.set_step(step);
+  summary_value = event.mutable_summary()->add_value();
+  summary_value->set_tag(dname + " Compute Time (ms)/" + cluster);
+  summary_value->set_simple_value((float)compute_time);
+  // done creating event object
+
+  if (files.empty())  // create new tf event file
+  {
+    str_path += ("stats" + to_string(file_idx));
+
+    tensorflow::EventsWriter writer(str_path);
+
+    try {
+      writer.WriteEvent(event);
+    } catch (...) {
+      return errors::Internal("Error writing event to new event file");
+    }
+  } else  // append to tf event file in dir
+  {
+    std::string file_path = str_path + files[0];
+
+    // open tf event file at location file_path
+    std::unique_ptr<tensorflow::WritableFile> writable_file;
+
+    env->NewAppendableFile(file_path, &writable_file);
+    tensorflow::io::RecordWriter record_writer(writable_file.get());
+    // done opening tf event file at location file_path
+
+    try {
+      record_writer.WriteRecord(event.SerializeAsString());
+    } catch (...) {
+      return errors::Internal("Error updating event file with new event");
+    }
+  }
+
+  return Status::OK();
+}
+
+Status CreateSummaryFromGraph(tensorflow::Graph* graph,
+                              std::string filename_prefix) {
+  // convert Graph* to GraphDef*
+  tensorflow::GraphDef* gdef = nullptr;
+  graph->ToGraphDef(gdef);
+
+  // create event summary writer
+  tensorflow::EventsWriter writer(filename_prefix);
+  tensorflow::Event event;
+
+  // write graph to event summary
+  try {
+    event.set_graph_def(gdef->SerializeAsString());
+    writer.WriteEvent(event);
+  } catch (...) {
+    return errors::Internal("Error writing graph to event file");
+  }
+
+  return Status::OK();
+}
+
+Status CreateSummaryFromGraphDef(tensorflow::GraphDef* graph_def,
+                                 std::string filename_prefix) {
+  // create event summary writer
+  tensorflow::EventsWriter writer(filename_prefix);
+  tensorflow::Event event;
+
+  // write graph to event summary
+  try {
+    event.set_graph_def(graph_def->SerializeAsString());
+    writer.WriteEvent(event);
+  } catch (...) {
+    return errors::Internal("Error writing graph def to event file");
+  }
+
+  return Status::OK();
+}
+
+void AddSessionNameAttr(int file_idx, std::set<string> nodes, Graph* graph) {
+  for (auto node : graph->op_nodes()) {
+    if (node->type_string() == "NGraphEncapsulate") {
+      node->AddAttr(("_session_name" + to_string(file_idx)),
+                    GetSessionName(file_idx, nodes));
+    }
+  }
+}
+
+std::string GetSessionName(int file_idx, std::set<std::string> nodes) {
+  std::string name = "";
+
+  if (nodes.size() == 0) {
+    name = to_string(file_idx);
+  } else {
+    std::string entry;
+
+    for (auto it = nodes.begin(); it != nodes.end();
+         ++it)  // get last element in set
+    {
+      entry = *it;
+    }
```

Comment:
I *think* you are trying to find the topologically last node here, yes?

sets do not have last elements. If the last node happens to be the last node of the graph, that is coincidental and we should not rely upon that behaviour.

If u want the last node, you could use toposort: https://github.com/tensorflow/ngraph-bridge/blob/76055645afbbe1a6041bf468937467e998f787fb/ngraph_bridge/ngraph_builder.cc#L4898

Also consider that graph could have 2 "last" nodes. like X<--A-->Y has 2 "last" nodes X and Y

Reply:
i tried with a few examples, it seemed that the last node always was the highest ordered node that was requested. i know sets dont have an ordering, but the order in which the elements were inserted into the set remain the same, correct? i was thinking this ordering is preserved (when the graph nodes are added to the set), and am simply selecting the last one. let me know if this logic makes sense

Reply:
It seems that we have a simple solution to this problem:

If you look in this file, there is a part where we do:

```
  std::set<string> fetch_nodes;
  for (const string& f : item.fetch) {
    int pos = f.find(":");
    fetch_nodes.insert(f.substr(0, pos));
  }```

Now `item.fetch` happens to be a vector. Which means, `fetch_nodes` in the above code could be a vector. That solves the problem of ordering.

It seems `GetSessionName`'s `nodes` argument is always called with fetch nodes. So we can rename it `fetch_nodes` in `GetSessionName`, to be explicit that we are passing a vector of fetch nodes and not all nodes.

Now, fetch nodes are the "last" nodes. they are "incomparable" in a partial ordering. I think if u pass a vector of fetch nodes, ALL of them should be used to create the string. But maybe I am misunderstanding this piece of code. Lets f2f tomorrow.

Reply:
picking random node in set as per discussion just now